### PR TITLE
#776 Use a central and unified scheduling service

### PR DESCRIPTION
**Pull request by Rentier**
**What's in the PR**
* Add new scheduling module
* Add new scheduler with backing executor service
* Refactor recommendation tasks to use new scheduler

I did not yet refactor the search tasks, as this needs capabilities from #773 and #774 .

**How to test manually**
* Use a project with recommender and see whether it still works.

**Automatic testing**
* [x] PR includes unit tests

**Documentation**
* [ ] PR updates documentation

### CODE REVIEW

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java [hunk 1]
```diff
@@ -310,7 +322,23 @@ public void onRecommenderDelete(RecommenderDeletedEvent aEvent)
...
+        // i.e. we do not start it here.
+        if (count % TRAININGS_PER_SELECTION == 0) {
+            Task task = new SelectionTask(aProject, user);
+            schedulingService.enqueue(task);
+        } else {
+            Task task = new TrainingTask(user, aProject);
+            schedulingService.enqueue(task);
+        }
+
+        trainingTaskCounter.put(key, count + 1);
```
Comment (reckart) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: I think incrementing the counter should be made an atomic synchronized operation.

File: inception-scheduling/src/main/java/de/tudarmstadt/ukp/inception/scheduling/IntrospectiveThreadPoolExecutor.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiConsumer;
+
+import org.apache.commons.lang3.concurrent.BasicThreadFactory;
+
+public class IntrospectiveThreadPoolExecutor
```
Comment (reckart) [inline inception-scheduling/src/main/java/de/tudarmstadt/ukp/inception/scheduling/IntrospectiveThreadPoolExecutor.java]: What does "introspective" mean here?

File: inception-scheduling/src/main/java/de/tudarmstadt/ukp/inception/scheduling/SchedulingService.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    private void afterExecute(Runnable aRunnable, Throwable aThrowable)
+    {
+        runningTasks.remove(aRunnable);
+    }
+
+    public List<Task> getScheduledTasks()
+    {
+        ArrayList<Task> result = new ArrayList<>();
```
Comment (reckart) [inline inception-scheduling/src/main/java/de/tudarmstadt/ukp/inception/scheduling/SchedulingService.java]: `List<Task> result = `

File: inception-scheduling/src/main/java/de/tudarmstadt/ukp/inception/scheduling/SchedulingService.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    public List<Task> getRunningTasks()
+    {
+        // We return copy here, as else the list the receiver sees might be updated
+        // when new tasks are running or existing ones stopped.
+        return new ArrayList<>(runningTasks);
+    }
+
+    public List<Task> getScheduledAndRunningTasks()
+    {
+        ArrayList<Task> result = new ArrayList<>();
```
Comment (reckart) [inline inception-scheduling/src/main/java/de/tudarmstadt/ukp/inception/scheduling/SchedulingService.java]: List<Task> result =

File: inception-scheduling/src/main/java/de/tudarmstadt/ukp/inception/scheduling/IntrospectiveThreadPoolExecutor.java (hunk 2 above)
Comment (Rentier) [reply inception-scheduling/src/main/java/de/tudarmstadt/ukp/inception/scheduling/IntrospectiveThreadPoolExecutor.java]: It means that you can look into what the executor does.

File: inception-scheduling/src/main/java/de/tudarmstadt/ukp/inception/scheduling/SchedulingService.java (hunk 4 above)
Comment (Rentier) [reply inception-scheduling/src/main/java/de/tudarmstadt/ukp/inception/scheduling/SchedulingService.java]: Done

File: inception-scheduling/src/main/java/de/tudarmstadt/ukp/inception/scheduling/IntrospectiveThreadPoolExecutor.java (hunk 2 above)
Comment (reckart) [reply inception-scheduling/src/main/java/de/tudarmstadt/ukp/inception/scheduling/IntrospectiveThreadPoolExecutor.java]: I think that would be "Inspectable" then. "Introspective" sounds like the executor looks inside itself.

**Comment by reckart** [comment]
Tried opening a document in one of my recommender playground projects and got this:

```
java.lang.NullPointerException: null
	at de.tudarmstadt.ukp.inception.recommendation.service.RecommendationServiceImpl.triggerTrainingAndClassification(RecommendationServiceImpl.java:335) ~[classes/:?]
	at de.tudarmstadt.ukp.inception.recommendation.service.RecommendationServiceImpl.afterAnnotationUpdate(RecommendationServiceImpl.java:304) ~[classes/:?]
	at de.tudarmstadt.ukp.inception.recommendation.service.RecommendationServiceImpl$$FastClassBySpringCGLIB$$7b2c1577.invoke(<generated>) ~[classes/:?]
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:684) ~[spring-aop-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at de.tudarmstadt.ukp.inception.recommendation.service.RecommendationServiceImpl$$EnhancerBySpringCGLIB$$a24f915d.afterAnnotationUpdate(<generated>) ~[classes/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.springframework.context.event.ApplicationListenerMethodAdapter.doInvoke(ApplicationListenerMethodAdapter.java:259) ~[spring-context-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.context.event.ApplicationListenerMethodAdapter.processEvent(ApplicationListenerMethodAdapter.java:179) ~[spring-context-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.context.event.ApplicationListenerMethodAdapter.onApplicationEvent(ApplicationListenerMethodAdapter.java:142) ~[spring-context-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172) ~[spring-context-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165) ~[spring-context-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139) ~[spring-context-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:398) ~[spring-context-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:355) ~[spring-context-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at de.tudarmstadt.ukp.clarin.webanno.api.dao.DocumentServiceImpl.writeAnnotationCas(DocumentServiceImpl.java:661) ~[classes/:?]
	at de.tudarmstadt.ukp.clarin.webanno.api.dao.DocumentServiceImpl$$FastClassBySpringCGLIB$$7e8e7e07.invoke(<generated>) ~[classes/:?]
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:749) ~[spring-aop-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:294) ~[spring-tx-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:98) ~[spring-tx-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.security.access.intercept.aopalliance.MethodSecurityInterceptor.invoke(MethodSecurityInterceptor.java:69) ~[spring-security-core-4.2.10.RELEASE.jar:4.2.10.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688) ~[spring-aop-5.1.3.RELEASE.jar:5.1.3.RELEASE]
	at de.tudarmstadt.ukp.clarin.webanno.api.dao.DocumentServiceImpl$$EnhancerBySpringCGLIB$$243359bb.writeAnnotationCas(<generated>) ~[classes/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.apache.wicket.proxy.LazyInitProxyFactory$JdkHandler.invoke(LazyInitProxyFactory.java:521) ~[wicket-ioc-8.2.0.jar:8.2.0]
	at com.sun.proxy.$Proxy183.writeAnnotationCas(Unknown Source) ~[?:?]
	at de.tudarmstadt.ukp.clarin.webanno.ui.annotation.AnnotationPage.actionLoadDocument(AnnotationPage.java:562) [classes/:?]
	at de.tudarmstadt.ukp.clarin.webanno.ui.annotation.AnnotationPage.actionLoadDocument(AnnotationPage.java:538) [classes/:?]
	at de.tudarmstadt.ukp.clarin.webanno.ui.annotation.AnnotationPage$1.onDocumentSelected(AnnotationPage.java:210) [classes/:?]
	at de.tudarmstadt.ukp.clarin.webanno.ui.annotation.dialog.OpenDocumentDialog$2.onClose(OpenDocumentDialog.java:102) [classes/:?]
	at org.apache.wicket.extensions.ajax.markup.html.modal.ModalWindow$WindowClosedBehavior.respond(ModalWindow.java:965) [wicket-extensions-8.2.0.jar:8.2.0]
...
```

**Comment by reckart** [comment]
This is the line:

```
        AtomicInteger count = trainingTaskCounter.putIfAbsent(key, new AtomicInteger(0));
```

I think you wanted to use `computeIfAbsent` here?

**Comment by Rentier** [comment]
I wanted to initialize the count to 0 if it is not set yet.

**Comment by Rentier** [comment]
I thought `putIfAbsent` returns the new value, it returns the old one (in that case `null`). You are right, one can use `computeIfAbsent` to fix it.

**Comment by reckart** [comment]
> I wanted to initialize the count to 0 if it is not set yet. I thought I tested that. I will look on Monday into it.

I know, but `putIfAbsent` does not return the value you expect (i.e. the new value if no value was set yet). It returns the previous value associated with the key.

**Comment by reckart** [comment]
Made the fix. Seems to be ok now.
